src/lib/y2users/config.rb

```diff
@@ -113,112 +119,65 @@ def merge!(other)
-    attr_reader :users_manager
+    # @return [UsersCollection]
+    attr_reader :users_collection
 
-    # Manager for groups
+    # Collection of groups
     #
-    # @return [ElementManager]
-    attr_reader :groups_manager
+    # @return [GroupsCollection]
+    attr_reader :groups_collection
 
-    # Attaches an user or group
+    # Attaches a user or group
     #
-    # An id is assigned to the given user/group, if needed.
+    # @raise [RuntimeError] if the element is already attached to a config
+    # @raise [RuntimeError] if an element with the same id already exists
     #
-    # @param element [User, Group]
+    # @param element [ConfigElement]
     def attach_element(element)
-      element.assign_internal_id(ElementId.next) if element.id.nil?
+      raise "Element #{element} is already attached to a config" if element.attached?
+      raise "Element #{element} already exists in this config" if exist_element?(element)
 
-      element.is_a?(User) ? users_manager.attach(element) : groups_manager.attach(element)
+      element.assign_config(self)
+
+      collection_for(element).add(element)
     end
 
-    # Detaches an user or group
+    # Detaches a user or group
+    #
+    # @raise [RuntimeError] if the given element is not attached to the config
     #
-    # @param element [User, Group]
+    # @param element [ConfigElement]
     def detach_element(element)
-      element.is_a?(User) ? users_manager.detach(element) : groups_manager.detach(element)
-    end
+      raise "Element #{element} is not attached to this config" if element.config != self
 
-    # Helper class to manage a list of users or groups
-    class ElementManager
-      include Yast::Logger
+      exist = exist_element?(element)
 
-      # @return [Array<User, Group>]
-      attr_reader :elements
-
-      # @return [Config]
-      attr_reader :config
-
-      # Constructor
-      #
-      # @param config [Config]
-      def initialize(config)
-        @config = config
-        @elements = []
+      if !exist
+        log.warn("Detach element: element #{element} is attached to the config #{self}, but " \
+          "it cannot be found.")
       end
 
-      # Attaches the element to the config
-      #
-      # @raise [RuntimeError] if the element is already attached
-      #
-      # @param element [User, Group]
-      def attach(element)
-        raise "Element #{element} already attached to config #{config}" if element.attached?
-
-        @elements << element
-
-        element.assign_config(config)
-      end
-
-      # Detaches the element from the config
-      #
-      # @raise [RuntimeError] if the given element is not attached to the config
-      #
-      # @param element [User, Group]
-      def detach(element)
-        raise "Element #{element} is not attached to config #{config}" if element.config != config
-
-        index = @elements.find_index { |e| e.is?(element) }
+      collection_for(element).delete(element.id) if exist
```

Comment:
In general, I don't like if/else if I don't need several lines of code for each block of the if/else. I prefer guard clauses. If fact, the `if exist` here could be suppressed, but I left it to avoid NOP calls.

Reply:
> In general, I don't like if/else if I don't need several lines of code for each block of the if/else. I prefer guard clauses.

I agree with you, but I suggested it here because it one of those case where they fit well: if exists, do that or log a warning if not. Much readable :)

Anyway, let's resolve the conversation since it was just a minor suggestion and, honestly, I have no strong preference here :)

Thanks!